File: app/src/main/java/com/keylesspalace/tusky/entity/Account.java
/* Copyright 2017 Andrew Dawson
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>. */

package com.keylesspalace.tusky.entity;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.Spanned;

import com.google.gson.annotations.SerializedName;
import com.keylesspalace.tusky.util.HtmlUtils;

public class Account implements Parcelable {
    public String id;

    @SerializedName("username")
    public String localUsername;

    @SerializedName("acct")
    public String username;

    @SerializedName("display_name")
    public String displayName;

    public Spanned note;

    public String url;

    public String avatar;

    public String header;

    public boolean locked;

    @SerializedName("followers_count")
    public String followersCount;

    @SerializedName("following_count")
    public String followingCount;

    @SerializedName("statuses_count")
    public String statusesCount;

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public boolean equals(Object other) {
        if (this.id == null) {
            return this == other;
        } else if (!(other instanceof Account)) {
            return false;
        }
        Account account = (Account) other;
        return account.id.equals(this.id);
    }

    public String getDisplayName() {
        if (displayName.length() == 0) {
            return localUsername;
        }
        return displayName;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(localUsername);
        dest.writeString(username);
        dest.writeString(displayName);
        dest.writeString(HtmlUtils.toHtml(note));
        dest.writeString(url);
        dest.writeString(avatar);
        dest.writeString(header);
        dest.writeBooleanArray(new boolean[] { locked });
        dest.writeString(followersCount);
        dest.writeString(followingCount);
        dest.writeString(statusesCount);
    }

    public Account() {}

    protected Account(Parcel in) {
        id = in.readString();
        localUsername = in.readString();
        username = in.readString();
        displayName = in.readString();
        note = HtmlUtils.fromHtml(in.readString());
        url = in.readString();
        avatar = in.readString();
        header = in.readString();
        boolean[] lockedArray = new boolean[1];
        in.readBooleanArray(lockedArray);
        locked = lockedArray[0];
        followersCount = in.readString();
        followingCount = in.readString();
        statusesCount = in.readString();
    }

    public static final Creator<Account> CREATOR = new Creator<Account>() {
        @Override
        public Account createFromParcel(Parcel source) {
            return new Account(source);
        }

        @Override
        public Account[] newArray(int size) {
            return new Account[size];
        }
    };
}

File: app/src/main/java/com/keylesspalace/tusky/entity/Account.kt
/* Copyright 2017 Andrew Dawson
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>. */

package com.keylesspalace.tusky.entity

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.util.Date

@JsonClass(generateAdapter = true)
data class Account(
    val id: String,
    @Json(name = "username") val localUsername: String,
    @Json( name = "acct") val username: String,
    // should never be null per Api definition, but some servers break the contract
    @Json(name = "display_name") val displayName: String? = null,
    @Json(name = "created_at") val createdAt: Date,
    val note: String,
    val url: String,
    val avatar: String,
    val header: String,
    val locked: Boolean = false,
    @Json(name = "followers_count") val followersCount: Int = 0,
    @Json(name = "following_count") val followingCount: Int = 0,
    @Json(name = "statuses_count") val statusesCount: Int = 0,
    val source: AccountSource? = null,
    val bot: Boolean = false,
    // default value for backward compatibility
    val emojis: List<Emoji> = emptyList(),
    // default value for backward compatibility
    val fields: List<Field> = emptyList(),
    val moved: Account? = null,
    val roles: List<Role> = emptyList()
) {

    val name: String
        get() = if (displayName.isNullOrEmpty()) {
            localUsername
        } else {
            displayName
        }

    val isRemote : Boolean
        get() = this.username != this.localUsername
}

@JsonClass(generateAdapter = true)
data class AccountSource(
    val privacy: Status.Visibility = Status.Visibility.PUBLIC,
    val sensitive: Boolean? = null,
    val note: String? = null,
    val fields: List<StringField> = emptyList(),
    val language: String? = null
)

@JsonClass(generateAdapter = true)
data class Field(
    val name: String,
    val value: String,
    @Json(name = "verified_at") val verifiedAt: Date? = null
)

@JsonClass(generateAdapter = true)
data class StringField(
    val name: String,
    val value: String
)

@JsonClass(generateAdapter = true)
data class Role(
    val name: String,
    val color: String
)
